shared/middleware/apps_api.rb

```diff
@@ -272,4 +273,39 @@ class AppsApi < Sinatra::Base
     call(env.merge('REQUEST_METHOD'=>'POST'))
   end
 
+  #
+  # POST /v3/apps/<app-id>/import-(shared|user)-tables/<table-name>
+  #
+  # Imports a csv form post into a table, erasing previous contents.
+  #
+  post %r{/v3/apps/([^/]+)/import-(shared|user)-tables/([^/]+)$} do |app_id, endpoint, table_name|
+    unsupported_media_type unless params[:import_file][:type]== 'text/csv'   
+    rows = CSV.parse(params[:import_file][:tempfile])
+    table = Table.new(app_id, storage_id(endpoint), table_name)
+    columns = rows[0]
+    columns.each_with_index do |column, i|
+          halt 400, {}, "The first row of the csv must contain the field names for your data. One of your field names is empty: " + columns.join(',') unless column
+    end
+    records = []
+    rows = rows[1..-1]
+    rows.each do |row|
+      record = {}
+      row.each_with_index do |value, i|
+        record[columns[i]] = value if value
+      end
+      records.push(record)
+    end
+
+    # wait for all records to be parsed before deleting the old records.
+    table.to_a.each do |row|
+      table.delete(row[:id])
+    end
+    
+    records.each do |record|
+      table.insert(record, request.ip)
+    end
+
+    redirect "/private/edit-csp-table/#{app_id}/#{table_name}"
+  end
+
```

Comment:
I think this can be simplified a lot:

```
post %r{/v3/apps/([^/]+)/import-(shared|user)-tables/([^/]+)$} do |app_id, endpoint, table_name|
  unsupported_media_type unless params[:import_file][:type] == 'text/csv'

  table = Table.new(app_id, storage_id(endpoint), table_name)

  begin
    CSV.foreach(params[:import_file][:tempfile], headers:true) do |row|
      records.push(row.to_hash)
    end
  rescue => e
    halt 400, {}, "Malformed CSV file (#{e.message})."
  end
  
  # wait for all records to be parsed before deleting the old records.
  table.to_a.each do |row|
    table.delete(row[:id])
  end
  
  records.each do |record|
    table.insert(record, request.ip)
  end

  redirect "/private/edit-csp-table/#{app_id}/#{table_name}"
end
```